Source/WebCore/rendering/RenderObject.cpp

```diff
@@ -2548,13 +2548,36 @@ static bool coalesceSelectionGeometryWithAdjacentQuadsIfPossible(SelectionGeomet
         return (first - second).diagonalLengthSquared() <= maxDistanceBetweenBoundaryPoints * maxDistanceBetweenBoundaryPoints;
     };
 
+    auto areOverlapping = [](const FloatQuad& first, const FloatQuad& second) {
+        bool intersects = false;
+
+        for (const auto& p : { first.p1(), first.p2(), first.p3(), first.p4() }) {
+            if (second.containsPoint(p))
+                intersects = true;
+        }
+
+        for (const auto& p : { second.p1(), second.p2(), second.p3(), second.p4() }) {
+            if (first.containsPoint(p))
+                intersects = true;
+        }
+
+        if (!intersects)
+            return false;
+
+        auto isTrailingSideOverlapping = (first.p1() - first.p2()).diagonalLength() + (second.p1() - second.p2()).diagonalLength() > (first.p1() - second.p2()).diagonalLength();
+        auto isLeadingSideOverlapping = (first.p3() - first.p4()).diagonalLength() + (second.p3() - second.p4()).diagonalLength() > (first.p3() - second.p4()).diagonalLength();
+
+        return isTrailingSideOverlapping && isLeadingSideOverlapping;
```

Comment:
Do both conditions have to be true for `areOverlapping`?

Reply:
I think this is the right logic for this particular scenario (i.e. "two adjacent selection rects are clipping into each other, in such a way that the top and bottom edges of both rects overlap").

But you raise a good point — perhaps, `areOverlapping` should renamed to something like `leadingAndTrailingEdgesOverlap`.

Reply:
yes, otherwise it may not be "properly" overlapping

Reply:
Ooh I get what's happening now. @rr-codes going to forward a minor optional suggestion from Wenson's comment - can you rename `areOverlapping` to something like `leadingAndTrailingEdgesOverlap`.